### PR TITLE
Added null-pointer guard around setChainParams() calls

### PR DESCRIPTION
**Pull request by bobsummerwill**

### CODE REVIEW

File: libethereum/ChainParams.cpp [hunk 1]
```diff
@@ -116,8 +116,9 @@ ChainParams ChainParams::loadGenesis(string const& _json, h256 const& _stateRoot
 SealEngineFace* ChainParams::createSealEngine()
 {
 	SealEngineFace* ret = SealEngineRegistrar::create(sealEngineName);
-	assert(ret && "Seal engine not found");
-	ret->setChainParams(*this);
+	assert((ret != nullptr) && "Seal engine not found");
+	if (ret != nullptr)
```
Comment (chriseth) [inline libethereum/ChainParams.cpp]: The coding style actually recommends `!ret` here.

Comment (bobsummerwill) [inline libethereum/ChainParams.cpp]: Why?   That is pretty poor advice, IMHO.
And where do you see that recommendation?

Comment (chriseth) [inline libethereum/ChainParams.cpp]: It is not explicit in the CodingStandards.txt, but it is used like that all over the codebase.
Pro: Much better readable, goes well with type changes
Con: Goes well with type changes

File: libethereum/ChainParams.cpp [hunk 2]
```diff
@@ -116,8 +116,9 @@ ChainParams ChainParams::loadGenesis(string const& _json, h256 const& _stateRoot
...
+		ret->setChainParams(*this);
 	if (sealRLP.empty())
 	{
 		sealFields = ret->sealFields();
```
Comment (chriseth) [inline libethereum/ChainParams.cpp:125]: You use ret again here, so probably better to return early or throw.

File: libethcore/SealEngine.cpp [hunk 3]
```diff
@@ -51,7 +51,8 @@ void SealEngineFace::verifyTransaction(ImportRequirements::value _ir, Transactio
 SealEngineFace* SealEngineRegistrar::create(ChainOperationParams const& _params)
 {
 	SealEngineFace* ret = create(_params.sealEngineName);
-	assert(!!ret && "Seal engine not found.");
-	ret->setChainParams(_params);
+	assert(ret && "Seal engine not found.");
```
Comment (chfast) [inline libethcore/SealEngine.cpp:54]: The assert is definitely a bad choice here if `create()` can fail.

File: libethcore/SealEngine.cpp [hunk 4]
```diff
@@ -51,7 +51,8 @@ void SealEngineFace::verifyTransaction(ImportRequirements::value _ir, Transactio
...
+	if (ret)
```
Comment (chfast) [inline libethcore/SealEngine.cpp:55]: I would prefer using "quick return" pattern as in other file.

File: libethereum/ChainParams.cpp [hunk 5]
```diff
@@ -117,6 +117,8 @@ SealEngineFace* ChainParams::createSealEngine()
 {
 	SealEngineFace* ret = SealEngineRegistrar::create(sealEngineName);
 	assert(ret && "Seal engine not found");
+	if (!ret)
```
Comment (chfast) [inline libethereum/ChainParams.cpp:120]: The same issue with assert.

File: libethereum/ChainParams.cpp [hunk 6]
```diff
@@ -117,6 +117,8 @@ SealEngineFace* ChainParams::createSealEngine()
...
+		return NULL;
```
Comment (chfast) [inline libethereum/ChainParams.cpp]: ~~NULL~~nullptr.

**Comment by bobsummerwill** [comment]
RE:   Quick return.

Original code prior to @chriseth assertion:

```
SealEngineFace* ret = create(_params.sealEngineName);
ret->setChainParams(_params);
return ret;
```

Current code:

```
SealEngineFace* ret = create(_params.sealEngineName);
assert(ret && "Seal engine not found.");
ret->setChainParams(_params);
return ret;
```

My proposed code:

```
SealEngineFace* ret = create(_params.sealEngineName);
assert(ret && "Seal engine not found.");
if (ret)
    ret->setChainParams(_params);
return ret;
```

So what are you suggesting, @chfast?   This?   Is it any better?

```
SealEngineFace* ret = create(_params.sealEngineName);
assert(ret && "Seal engine not found.");
if (!ret)
    return NULL;
ret->setChainParams(_params);
return ret;
```

**Comment by chfast** [comment]
In general I propose to remove the asserts. The rest is about coding style only.

**Comment by bobsummerwill** [comment]
NOTE - So we had nothing.    Then we added an assert for better error messages in debug builds.   And now I am attempting to NOT crash.      If we really want to follow through here, we should add handling in the sites which call this too.    I've not looked at all those yet.    I'm just trying to improve the code we just modified, so we can get at least these functions in a state which we like.

**Comment by bobsummerwill** [comment]
OK - over to you @chriseth :-)

**Comment by chfast** [comment]
If asserts are there temporary for diagnostics then add a comment about that and leave them. And add an issue to track the bug later one.

**Comment by chriseth** [comment]
So in order to fix this properly, we need to check the call sites.
I would also prefer to add an assert to guard against future sloppy usage.
This is not about finding the cause of a problem but rather about failing fast if the problem occurs or persists.

**Comment by bobsummerwill** [comment]
Yes - having the assert AND the runtime check is not a bad thing.

Can we get this committed as-is, please, @chriseth?    It will make things better.

**Comment by chfast** [comment]
Just replace `NULL` with `nullptr`.

**Comment by bobsummerwill** [comment]
DONE

**Comment by chfast** [comment]
So I'm ok with the change. Not sure if you need more confirmations.

**Pull request merged**